powershell/gen_powershell_binding.ml

```diff
@@ -287,40 +305,38 @@ and print_converters classes =
-    let removers = fst partRem in
-    let partAdd = List.partition (fun x -> is_adder x) (snd partRem) in
-    let adders = fst partAdd in
-    let partSet = List.partition (fun x -> is_setter x) (snd partAdd) in
-    let setters = fst partSet in
-    let partGet = List.partition (fun x -> is_getter x) (snd partSet) in
-    let getters = fst partGet in
-    let invokers = List.filter (fun x -> is_invoke x) (snd partGet) in
-    let stem = ocaml_class_to_csharp_class classname in
-    begin
-      if List.mem classname classes_with_records then
-        write_file (sprintf "Get-Xen%s.cs" stem) (gen_class obj classname);
-      if ((List.length constructors) > 0) then
-        write_file (sprintf "New-Xen%s.cs" stem) (gen_constructor obj classname constructors);
-      if ((List.length destructors) > 0) then
-        write_file (sprintf "Remove-Xen%s.cs" stem) (gen_destructor obj classname destructors);
-      if ((List.length removers) > 0) then
-        write_file (sprintf "Remove-Xen%sProperty.cs" stem) (gen_remover obj classname removers);
-      if ((List.length adders) > 0) then
-        write_file (sprintf "Add-Xen%s.cs" stem) (gen_adder obj classname adders);
-      if ((List.length setters) > 0) then
-        write_file (sprintf "Set-Xen%s.cs" stem) (gen_setter obj classname setters);
-      if ((List.length getters) > 0) then
-        write_file (sprintf "Get-Xen%sProperty.cs" stem) (gen_getter obj classname getters);
-      if ((List.length invokers) > 0) then
-        write_file (sprintf "Invoke-Xen%s.cs" stem) (gen_invoker obj classname invokers);
-    end
+  let {name=classname; messages} = obj in
+  let stem = ocaml_class_to_csharp_class classname in
+
+  let cmdlets = [{
+      filename = sprintf "Get-Xen%s.cs" stem;
+      content = gen_class obj classname;
+    }; {
+      filename = sprintf "New-Xen%s.cs" stem;
+      content = gen_constructor obj classname (List.filter is_constructor messages)
+    }; {
+      filename = sprintf "Remove-Xen%s.cs" stem;
+      content = gen_destructor obj classname (List.filter is_destructor messages);
+    }; {
+      filename = sprintf "Remove-Xen%sProperty.cs" stem;
+      content = gen_remover obj classname (List.filter is_remover messages);
+    }; {
+      filename = sprintf "Add-Xen%s.cs" stem;
+      content = gen_adder obj classname (List.filter is_adder messages);
+    }; {
+      filename = sprintf "Set-Xen%s.cs" stem;
+      content = gen_setter obj classname (List.filter is_setter messages);
+    }; {
+      filename = sprintf "Get-Xen%sProperty.cs" stem;
+      content = gen_getter obj classname (List.filter is_getter messages);
+    }; {
+      filename = sprintf "Invoke-Xen%s.cs" stem;
+      content = gen_invoker obj classname (List.filter is_invoke messages);
+    };
+  ] in
+
+  let filtered = List.filter (fun x -> x.content <> "") cmdlets in
+  List.map (fun x-> write_file x.filename x.content; x.filename) filtered;
```

Comment:
A better name would be `non_empty` or similar. I would also consider doing filter and map-ing in one go:
```
cmdlets
|> List.filter (fun x -> x.content <> "")
|> List.map ...
```

Is it correct that `List.map` has a side effect (writing a file) and returns values? Otherwise `List.iter` would be better. I would also consider creating a function to handle writing to a file.

```
let with_output filename f =
  let io = open_out filename in
  finally 
    (fun () -> f io)
    (fun () -> close_out io)
```

Reply:
Yes, it does look weird; I've used List.map because I needed somehow to collect the file names together in order to populate the project file in lines 105-114. I'm happy to change it if a cleaner way to achieve this can be suggested.

write_file is defined in line 341 - do you mean creating one function in the common folder of this repo? (Indeed each generator file in this repo defines its own).

Reply:
I like a function like `with_output` because opening, closing, and exception handling are in one place. 

I don't think it matters much but I would try to compute content and filenames first such that I would not have to rely on the function emitting the content to return the name if you need it. It's not worth spending too much energy on this. The code is clear in general.

Reply:
Thanks, I think it looks better now.
I've only corrected this one instance of `write_file` to use `with_output`; I'll do the other instances in this repo in a different PR, because I would like to avoid conflicts in view of the upcoming merge of the `json` branch into master.